org.osgi.test.junit5.cm/src/test/java/org/osgi/test/junit5/cm/ConfigAnnotationTest.java

```diff
@@ -0,0 +1,182 @@
+
+	@BeforeAll
+	public static void beforeAll() {
+
+		bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+
+		sref = bundleContext.getServiceReference(ConfigurationAdmin.class);
+		if (sref == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+		ca = bundleContext.getService(sref);
+		if (ca == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+	}
+
+	@AfterAll
+	public static void afterAll() {
+		BundleContext bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+		if (bundleContext != null && sref != null) {
+			bundleContext.ungetService(sref);
+		}
+
+	}
+
+	// START TESTS
+	@InjectConfiguration(NONSTATIC_CONFIGURATION_PID)
+	Configuration nonStaticConfiguration;
+
+	@Test
+	public void testFieldConfiguration() throws Exception {
+
+		Configuration cs = ConfigUtil.getConfigsByServicePid(ca, NONSTATIC_CONFIGURATION_PID);
+
+		assertThat(cs).isEqualTo(nonStaticConfiguration);
+
+	}
+
+	static final String PARAM_PID = "param.pid";
+
+	@Test
+	@WithConfiguration(pid = PARAM_PID, properties = {
+		@ConfigEntry(key = "bar", value = "foo")
+	})
+	public void testParameterConfiguration(
+		@InjectConfiguration(PARAM_PID) Configuration configuration)
+		throws Exception {
+
+		Configuration cs = ConfigUtil.getConfigsByServicePid(ca, PARAM_PID);
+		assertThat(cs).isEqualTo(configuration);
+
+		assertNull(configuration.getProperties()
+			.get("foo"));
+
+		String foo = (String) configuration.getProperties()
+			.get("bar");
+		assertEquals("foo", foo);
+	}
+
+	@Test
+	@WithConfiguration(pid = PARAM_PID, properties = {
+		@ConfigEntry(key = "foo", value = "bar")
+	})
+	public void testParameterConfiguration2(
+		@InjectConfiguration(PARAM_PID) Configuration configuration)
+		throws Exception {
+
+		Configuration cs = ConfigUtil.getConfigsByServicePid(ca, PARAM_PID);
+		assertThat(cs).isEqualTo(configuration);
+
+		assertNull(configuration.getProperties()
+			.get("bar"));
+		String bar = (String) configuration.getProperties()
+			.get("foo");
+		assertEquals("bar", bar);
+	}
+
+	@Test
+	@WithConfiguration(pid = PARAM_PID)
+	public void testParameterConfiguration3(
+		@InjectConfiguration(PARAM_PID) Configuration configuration)
+		throws Exception {
+
+		Configuration cs = ConfigUtil.getConfigsByServicePid(ca, PARAM_PID);
+
+		assertNull(configuration.getProperties()
+			.get("foo"));
+		assertNull(configuration.getProperties()
+			.get("bar"));
+	}
+
+	static final String METHOD_PID = "method.pid";
+
+	@Test
+	@WithConfiguration(pid = METHOD_PID, properties = {
+		@ConfigEntry(key = "foo", value = "bar")
+	})
+	public void testMethodConfiguration() throws Exception {
+
+		Configuration cs = ConfigUtil.getConfigsByServicePid(ca, METHOD_PID);
+		assertThat(cs).isNotNull();
+		assertEquals("bar", cs.getProperties()
+			.get("foo"));
```

Comment:
This is much nicer as:

```java
assertThat(Dictionaries.asMap(configuration.getProperties())).containsEntry("foo", "bar");
```

One of the reasons I created Dictionaries.asMap was for assertj testing :-)

Reply:
We should have our own custom entry point (soft and hard) for `assertThat(Dictionary)`, as this would make this nicer still. I've noticed it's a common pattern.

Well, I went the extra step:
https://github.com/osgi/osgi-test/pull/174

Reply:
done